From b5ce739ded823a58720b6d031dbf7aeda161a9c1 Mon Sep 17 00:00:00 2001
From: Ilies CHERGUI <ichergui@nvidia.com>
Date: Thu, 15 Jan 2026 12:47:09 -0800
Subject: [PATCH 08/11] Fix unifiedgpudisp modules builds

Update the makefiles, Kbuild, and conftest.sh script so
that headers and symbols from the out-of-tree drivers can
be found when building the display driver modules.

Upstream-Status: Inappropriate [OE specific]

Signed-off-by: Ilies CHERGUI <ichergui@nvidia.com>
---
 kernel-open/Kbuild      | 15 ++++++++-------
 kernel-open/Makefile    |  1 +
 kernel-open/conftest.sh | 32 ++++++++++++++++++--------------
 3 files changed, 27 insertions(+), 21 deletions(-)

diff --git a/kernel-open/Kbuild b/kernel-open/Kbuild
index 710570f..9ade365 100644
--- a/kernel-open/Kbuild
+++ b/kernel-open/Kbuild
@@ -77,6 +77,7 @@ $(foreach _module, $(NV_KERNEL_MODULES), \
 
 ccflags-y += -I$(src)/common/inc
 ccflags-y += -I$(src)
+ccflags-y += -I$(NV_OOT_SOURCES)/include
 ccflags-y += -Wall $(DEFINES) $(INCLUDES) -Wno-cast-qual -Wno-format-extra-args
 ccflags-y += -D__KERNEL__ -DMODULE -DNVRM
 ccflags-y += -DNV_VERSION_STRING=\"580.00\"
@@ -178,7 +179,7 @@ NV_CONFTEST_SCRIPT := $(src)/conftest.sh
 NV_CONFTEST_HEADER := $(obj)/conftest/headers.h
 
 NV_CONFTEST_CMD := /bin/sh $(NV_CONFTEST_SCRIPT) \
- "$(CC)" $(ARCH) $(NV_KERNEL_SOURCES) $(NV_KERNEL_OUTPUT)
+ "$(CC)" $(ARCH) $(NV_KERNEL_SOURCES) $(NV_KERNEL_OUTPUT) $(NV_OOT_SOURCES)
 
 NV_CFLAGS_FROM_CONFTEST := $(shell $(NV_CONFTEST_CMD) build_cflags)
 
@@ -244,8 +245,8 @@ $(eval $(call NV_GENERATE_COMPILE_TEST_HEADER,symbols,$(NV_CONFTEST_SYMBOL_COMPI
 $(eval $(call NV_GENERATE_COMPILE_TEST_HEADER,types,$(NV_CONFTEST_TYPE_COMPILE_TESTS)))
 
 $(obj)/conftest/patches.h: $(NV_CONFTEST_SCRIPT)
-	@mkdir -p $(obj)/conftest
-	@$(NV_CONFTEST_CMD) patch_check > $@
+	mkdir -p $(obj)/conftest
+	$(NV_CONFTEST_CMD) patch_check > $@
 
 include $(src)/header-presence-tests.mk
 
@@ -256,8 +257,8 @@ NV_HEADER_PRESENCE_PART = $(addprefix $(obj)/conftest/header_presence/,$(addsuff
 # Define a rule to check the header $(1).
 define NV_HEADER_PRESENCE_CHECK
  $$(call NV_HEADER_PRESENCE_PART,$(1)): $$(NV_CONFTEST_SCRIPT) $(obj)/conftest/uts_release
-	@mkdir -p $$(dir $$@)
-	@$$(NV_CONFTEST_CMD) test_kernel_header '$$(NV_CONFTEST_CFLAGS)' '$(1)' > $$@
+	mkdir -p $$(dir $$@)
+	$$(NV_CONFTEST_CMD) test_kernel_header '$$(NV_CONFTEST_CFLAGS)' '$(1)' > $$@
 endef
 
 # Evaluate the rule above for each header in the list.
@@ -265,7 +266,7 @@ $(foreach header,$(NV_HEADER_PRESENCE_TESTS),$(eval $(call NV_HEADER_PRESENCE_CH
 
 # Concatenate all of the parts into headers.h.
 $(obj)/conftest/headers.h: $(call NV_HEADER_PRESENCE_PART,$(NV_HEADER_PRESENCE_TESTS))
-	@cat $^ > $@
+	cat $^ > $@
 
 clean-dirs := $(obj)/conftest
 
@@ -287,7 +288,7 @@ BUILD_SANITY_CHECKS = \
 .PHONY: $(BUILD_SANITY_CHECKS)
 
 $(BUILD_SANITY_CHECKS):
-	@$(NV_CONFTEST_CMD) $@ full_output
+	$(NV_CONFTEST_CMD) $@ full_output
 
 # Perform all sanity checks before generating the conftest headers
 
diff --git a/kernel-open/Makefile b/kernel-open/Makefile
index 245d91b..ea53d57 100644
--- a/kernel-open/Makefile
+++ b/kernel-open/Makefile
@@ -129,6 +129,7 @@ else
   KBUILD_PARAMS += ARCH=$(ARCH)
   KBUILD_PARAMS += NV_KERNEL_SOURCES=$(KERNEL_SOURCES)
   KBUILD_PARAMS += NV_KERNEL_OUTPUT=$(KERNEL_OUTPUT)
+  KBUILD_PARAMS += NV_OOT_SOURCES=$(OOTSRC)
   KBUILD_PARAMS += NV_KERNEL_MODULES="$(NV_KERNEL_MODULES)"
   KBUILD_PARAMS += INSTALL_MOD_DIR="$(INSTALL_MOD_DIR)"
   KBUILD_PARAMS += NV_SPECTRE_V2=$(SPECTRE_V2_RETPOLINE)
diff --git a/kernel-open/conftest.sh b/kernel-open/conftest.sh
index c4171a0..9c34e18 100755
--- a/kernel-open/conftest.sh
+++ b/kernel-open/conftest.sh
@@ -11,6 +11,8 @@ ARCH=$2
 SOURCES=$3
 HEADERS=$SOURCES/include
 OUTPUT=$4
+OOT_SOURCES=$5
+OOT_HEADERS=$OOT_SOURCES/include
 XEN_PRESENT=1
 PREEMPT_RT_PRESENT=0
 
@@ -172,6 +174,8 @@ build_cflags() {
     CFLAGS="$CFLAGS -I$SOURCE_ARCH_HEADERS/uapi"
     CFLAGS="$CFLAGS -I$OUTPUT_ARCH_HEADERS/generated"
     CFLAGS="$CFLAGS -I$OUTPUT_ARCH_HEADERS/generated/uapi"
+    CFLAGS="$CFLAGS -I$OOT_HEADERS"
+    CFLAGS="$CFLAGS -I$OOT_HEADERS/uapi"
 
     if [ -n "$BUILD_PARAMS" ]; then
         CFLAGS="$CFLAGS -D$BUILD_PARAMS"
@@ -341,7 +345,7 @@ check_symbol_exists() {
         # Check Module.symvers to see whether the given symbol is present.
         #
         if grep -e "${TAB}${SYMBOL}${TAB}.*${TAB}EXPORT_SYMBOL.*\$" \
-                    ${MODULE_SYMVERS_PATHS} >/dev/null 2>&1; then
+                    ${MODULE_SYMVERS_PATHS} "$OOT_SOURCES/Module.symvers" >/dev/null 2>&1; then
             return 0
         fi
     else
@@ -399,7 +403,7 @@ export_symbol_gpl_conftest() {
     TAB='	'
 
     if grep -e "${TAB}${SYMBOL}${TAB}.*${TAB}EXPORT_\(UNUSED_\)*SYMBOL_GPL\s*\$" \
-                ${MODULE_SYMVERS_PATHS} >/dev/null 2>&1; then
+                ${MODULE_SYMVERS_PATHS} "$OOT_SOURCES/Module.symvers" >/dev/null 2>&1; then
         echo "#define NV_IS_EXPORT_SYMBOL_GPL_$SYMBOL 1" |
             append_conftest "symbols"
     else
@@ -5083,13 +5087,13 @@ compile_test() {
     esac
 }
 
-case "$5" in
+case "$6" in
     cc_sanity_check)
         #
         # Check if the selected compiler can create object files
         # in the current environment.
         #
-        VERBOSE=$6
+        VERBOSE=$7
 
         echo "int cc_sanity_check(void) {
             return 0;
@@ -5156,7 +5160,7 @@ case "$5" in
         #  In order to extract GCC version correctly for version strings
         #  like the last one above, we first check for x.y.z and if that
         #  fails, we fallback to x.y format.
-        VERBOSE=$6
+        VERBOSE=$7
 
         kernel_compile_h=$OUTPUT/include/generated/compile.h
 
@@ -5235,7 +5239,7 @@ case "$5" in
         # Check if the target kernel is a Xen kernel. If so, exit, since
         # the RM doesn't currently support Xen.
         #
-        VERBOSE=$6
+        VERBOSE=$7
 
         if [ -n "$IGNORE_XEN_PRESENCE" -o -n "$VGX_BUILD" ]; then
             exit 0
@@ -5267,7 +5271,7 @@ case "$5" in
         # Check if the target kernel has the PREEMPT_RT patch set applied. If
         # so, exit, since the RM doesn't support this configuration.
         #
-        VERBOSE=$6
+        VERBOSE=$7
 
         if [ -n "$IGNORE_PREEMPT_RT_PRESENCE" ]; then
             exit 0
@@ -5330,7 +5334,7 @@ case "$5" in
         # Run a series of compile tests to determine the set of interfaces
         # and features available in the target kernel.
         #
-        shift 5
+        shift 6
 
         CFLAGS=$1
         shift
@@ -5344,7 +5348,7 @@ case "$5" in
         #
         # Determine whether running in DOM0.
         #
-        VERBOSE=$6
+        VERBOSE=$7
 
         if [ -n "$VGX_BUILD" ]; then
             if [ -f /proc/xen/capabilities ]; then
@@ -5366,7 +5370,7 @@ case "$5" in
         #
         # Determine whether we are running a vGPU on KVM host.
         #
-        VERBOSE=$6
+        VERBOSE=$7
         iommu=CONFIG_VFIO_IOMMU_TYPE1
         iommufd_vfio_container=CONFIG_IOMMUFD_VFIO_CONTAINER
         mdev=CONFIG_VFIO_MDEV
@@ -5436,7 +5440,7 @@ case "$5" in
         #
         # Check to see if the given config option is set.
         #
-        OPTION=$6
+        OPTION=$7
 
         test_configuration_option $OPTION
         exit $?
@@ -5446,7 +5450,7 @@ case "$5" in
         #
         # Get the value of the given config option.
         #
-        OPTION=$6
+        OPTION=$7
 
         get_configuration_option $OPTION
         exit $?
@@ -5481,9 +5485,9 @@ case "$5" in
         # Check for the availability of the given kernel header
         #
 
-        CFLAGS=$6
+        CFLAGS=$7
 
-        test_header_presence "${7}"
+        test_header_presence "${8}"
 
         exit $?
     ;;
-- 
2.43.0

